### ISSUE TITLE
Another question about v5 concurrency?

### ISSUE THREAD

**Issue by kuiperzone** [litedb-org/LiteDB#1349]
I seem to recall seeing in the description that there are no read-locks in V5.

Does this mean that it will be possible to safely have one writer and one or more readers concurrently accessing the same DB?

Thanks

**Comment by kuiperzone** [comment]
Just to add a little information to my question, under the V5 description, it says:

"No lock for reader"

But, as you have confirmed, the file will be opened in exclusive mode. Does this apply to when a file is open in "ReadOnly" mode?

**Comment by kuiperzone** [comment]
Also, in the v5 connection string the Mode seems to be one of Shared or Embedded.

What will Shared do? And what's the difference?

Cheers

**Comment by mbdavid** [comment]
Hi @kuiperzone,

v5 works with snapshots - this means writer transaction has an insolated data map during writing. Read threads still reading old data until writer transaction commit. In v4, when writer transaction starts, no ones can read anymore.

Datafile are open with a single writer and many readers (stream pool). But only 1 writer is possible.

Embedded more are default mode: only 1 instance of LiteEngine are possible.

Shared mode are not 100% ok, i'm still working on it. But the idea here is keep datafile disconnected to be possible has many instances of LiteDatabase but working with only 1 single LiteEngine open.

**Comment by kuiperzone** [comment]
I guess I'm confused because you are thinking of concurrency in terms of threads calling on the same LiteDatabase instance, whereas I am thinking of "many readers" in terms of different processes and/or different instances of LiteDatabase.

However, I think I understand -- there is to be no concurrency other than that of threads (given that there does not seem to be a "ReadOnly" mode in v5).

Again, thank you for taking the time to reply. And again, I think you approach is probably the best one for LiteDB given that it is not client-server but a local solution.

**Comment by kuiperzone** [comment]
>Embedded more are default mode: only 1 instance of LiteEngine are possible.

Will Embedded mode mean that when reading a file, access will be faster provided there is no other thread writing?

Presumably, it will still possible have different instances of "LiteDatabase" accessing different database files in different threads?

**Comment by mbdavid** [comment]
Hi @kuiperzone, sorry for confusion. It's so native to me that I forgot that's not to easy :)

Let's me try again:

Embeded Mode (default) - Only a single instance of LiteEngine (or LiteDatabase|LiteRepository). Over this single instance you can run multiple threads.
- You can read N concurrent read transactions (one transaction PER thread). There is not lock for read operations. If needed, LiteDB use snapshot to read "old data version" when you are changing in another thread. LiteDB can use multiples `FileStream` in readonly mode.

- You can write N concurrent transactions (one transaction PER thread) - but only 1 write transaction PER collection. If you try write with 2 transaction in same collection, the second one will be locked until first transaction release (commit|rollback). LiteDB use another single writer queue thread to write on disk... using a single writeable `FileStream`.

In this mode, you can't have 2 instances of LiteEngine (or LiteDatabase|LiteRepository) because engine contains in-memory caches that are not shared between processes.

Shared Mode (slow and not great for production) - This is just a layer for LiteDatabase support multiple processes using same datafile. Using `Mutex`, it's just a simple: only one can access database per time. It´s not recommended because all optimizations are not used (like cache) and open/close datafile are expensive. You can take a look on this class, you will see only "Open/Execute/Close":

https://github.com/mbdavid/LiteDB/blob/master/LiteDB/Client/Shared/SharedEngine.cs

**Comment by mbdavid** [comment]
Yes, per datafile! Each instance contains its own cache. There is no static singleton.

**Comment by kuiperzone** [comment]
Brilliant. :)

**Comment by nhaberl** [comment]
So in case of asp.net core app we should handle each LiteDatabase instance as singleton per physical database file?

How would the architecture be when having 1 central db and each for every user ?

**Comment by TigerHix** [comment]
Can confirm multi-threading works after switching to Embedded Mode. Apparently Shared Mode's mutex would either be broken or deadlocked if r/w from multiple threads... This should be clarified in the wiki! :)

**Comment by I-Vladimir** [comment]
Working on an implementation of a small XML database I ran into the same issue. But I managed to resolve that using file locks beyond the end of file. I gave priority to readers, and assume my starting lock offset is 1000 and assume max number of reader equal to number of threads and also say equals to 1000. Then to read thread #1 can lock single byte offset 1001, thread #2 locks 1002. To write - to lock 1000 bytes offset starting from 1001. In such case if no writers - readers can lock its own byte with no concurrency.

**Comment by JKamsker** [comment]
Here's how it works in v5 (5.0.21; the 6.x prereleases behave the same):

- **Direct mode (default).** Use one `LiteDatabase` per data file per process and share it between all your threads. Reads take no lock: a reader on one thread still sees the last committed data while another thread's write transaction is open, and sees the new data after `Commit()`. Writes are serialized per collection. If you open a second Direct instance on the same file, it fails with an `IOException` ("file is being used by another process"). Instances on *different* files are independent, each with its own cache, and there is no global singleton.
- **Shared mode** (`Connection=Shared`) is for several processes on one file. Every operation takes a named mutex, opens the file, runs and closes it again. That makes it much slower and it has no cache.
- **`ReadOnly=true`** exists in the connection string, but it doesn't give you a live cross-process reader. It opens alongside a writer, but it does not see commits made after it was opened, and writes through it fail. For a reader in another process, use Shared mode.

```csharp
// one instance per file, shared by every thread in the process
var db = new LiteDatabase("Filename=app.db;Connection=Direct");

// only if several processes must open the same file
var shared = new LiteDatabase("Filename=app.db;Connection=Shared");
```

nhaberl: for ASP.NET Core, register one instance per database file as a singleton (details in #1412).

Closing this as answered — if it doesn't cover your case, comment here and we'll reopen it.
